vss_collection.py

```diff
@@ -0,0 +1,183 @@
+        range_name = _DEFAULT_RANGE_NAME
+        if isinstance(instance, str):
+            if re.match(reg_ex, instance):
+                inst_range_arr = re.split(r"\[+|,+|\]", instance)
+                range_name = inst_range_arr[0]
+                lower_bound = int(inst_range_arr[1])
+                upper_bound = int(inst_range_arr[2])
+                for element in range(lower_bound, upper_bound + 1):
+                    result.append(f"{range_name}{element}")
+
+                return VssInstance(range_name, result, True)
+
+            raise ValueError("", "", f"instantiation type {instance} not supported")
+
+        # Use list elements for instances (e.g. ["LEFT","RIGHT"])
+        if isinstance(instance, list):
+            for element in instance:
+                result.append(f"{element}")
+            return VssInstance(range_name, result, False)
+
+        raise ValueError("", "", f"is of type {type(instance)} which is unsupported")
+
+    def __gen_collection_types(self, name, instances) -> tuple[str, list]:
+        _type_name = f"{name}{_TYPE_SUFFIX}"
+        result = list()
+        print(f"{' ' * 5}- {_type_name:25}{instances}")
+        result.append(f"{self.ctx.tab}class {_type_name}(Model):")
+        result.append(f"{self.ctx.tab * 2}def __init__(self, name, parent):")
+        result.append(f"{self.ctx.tab * 3}super().__init__(parent)")
+        result.append(f"{self.ctx.tab * 3}self.name = name")
+
+        for instance in instances:
+            result.append(
+                f'{self.ctx.tab * 3}self.{instance} = {name}("{instance}", self)'
+            )
+
+        return (_type_name, result)
+
+    def __gen_getter(self, name, instances, indent_level) -> List[str]:
+        result = list()
+        count = len(instances)
+        result.append(f"\n{self.ctx.tab * indent_level}def {name}(self, index: int):")
+        result.append(
+            f"{self.ctx.tab * (indent_level + 1)}if index < 1 or index > {count}:"
+        )
+        result.append(
+            f'{self.ctx.tab * (indent_level + 2)}raise IndexError(f"Index {{index}} is out of range")'
+        )
+
+        result.append(f"{self.ctx.tab * (indent_level + 1)}_options = {{")
+        for i in range(len(instances)):
+            result.append(
+                f"{self.ctx.tab * (indent_level + 2)}{i + 1} : self.{instances[i]},"
+            )
+        result.append(f"{self.ctx.tab * (indent_level + 1)}}}")
+
+        result.append(
+            f"{self.ctx.tab * (indent_level + 1)}return _options.get(index)\n"
+        )
+        return result
+
+    def __gen_collection(self, node: VSSNode) -> List[str]:
+        print(f"- {self.name:30}{node.instances}")
+        result = list()
+        result.append(f"class {self.name}(Model):")
+        result.append(f"{self.ctx.tab}def __init__(self, name, parent):")
+        result.append(f"{self.ctx.tab * 2}super().__init__(parent)")
+        result.append(f"{self.ctx.tab * 2}self.name = name")
```

Comment:
why don't you use the indent-function or even better `with ctx as indented_context`. makes the readability much better. Have a look to the cpp-generator for usage-details.

Reply:
indent is only used when directly write to the generated files which allows the writer to increase the indentation level on the current context element. Due to the way we create the collector in a list elements makes it easier to append the necessary tab directly to the string.

Reply:
Fixed as discussed!